### PR TITLE
Save filesize with attachment metadata on upload

### LINKED ISSUE #492: Idea: store filesize attachment meta value on upload.
There are a few places in core, like `wp_prepare_attachment_for_js`, that do a filesystem stat check for attachments to get the filesize in bytes:

https://github.com/WordPress/WordPress/blob/master/wp-includes/media.php#L3969-L3975

This can be short-circuited if the "filesize" property is set in the attachment meta array.

Setting this property seems as though it would eliminate the need for several AWS S3 requests which happen as part of the `query-attachments` ajax action.  Is it feasible to set it on uploading attachments?

### PR DESCRIPTION
**Pull request by goldenapples**
Sets the "filesize" property in the attachment metadata when initially uploading images, so that it doesn't have to be queried for in the ajax response on `wp_prepare_attachment_for_js()`.

Fixes #492.

### CODE REVIEW

**Comment by joehoyle** [comment]
Nice, I had assumed core already did this -- do you know why it doesn't?

File: inc/class-plugin.php [hunk 1]
```diff
@@ -396,6 +397,33 @@ public function filter_sideload_move_temp_file_to_s3( array $file ) {
...
+	 * longer to complete against s3 uploads than if they were performed with a
+	 * local filesystem.i
+	 *
+	 * Saving the filesize in the attachment metadata when the image is
+	 * uploaded allows core to skip this stat when retrieving and formatting it.
+	 *
+	 * @param array $metadata Attachment metadata.
+	 * @return array Attachment metadata array, with "filesize" value added.
+	 */
+	function set_filesize_in_attachment_meta( $metadata ) {
```
Comment (joehoyle) [inline inc/class-plugin.php:415]:
```suggestion
	function set_filesize_in_attachment_meta( array $metadata ) : array {
```

**Comment by joehoyle** [comment]
The unit tests are failing here, because `metadata['file']`, surprisingly is not always set. It seems that's only for images, videos for example never get that set (https://core.trac.wordpress.org/browser/tags/5.6.1/src/wp-admin/includes/media.php#L3532). It's possible to get the file instead with `get_attached_file()` rather than reading from the metadata for that value. TIL

File: inc/class-plugin.php (hunk 1 above)
Comment (joehoyle) [reply inc/class-plugin.php:415]: This project is using Psalm too, so once the unit tests are fixed, I'd imagine we'll need to add the array shape decelerating like elsewhere in the file.

**Comment by goldenapples** [comment]
> I had assumed core already did this -- do you know why it doesn't?

I don't think there was a reason it wasn't added. The logic to check for "filesize" in attachment meta was added in 4.4 ([ticket](https://core.trac.wordpress.org/ticket/33214)) to support external image libraries and I guess the assumption was that those plugins had to implement attachment metadata themselves. Looks like there's a ticket open to set filesize in this metadata array in core - https://core.trac.wordpress.org/ticket/49412 - and if that gets merged this workaround can be removed.

File: inc/class-plugin.php [hunk 2]
```diff
@@ -396,6 +397,32 @@ public function filter_sideload_move_temp_file_to_s3( array $file ) {
...
+	 * uploaded allows core to skip this stat when retrieving and formatting it.
+	 *
+	 * @param array{file?: string} $metadata      Attachment metadata.
+	 * @param int                  $attachment_id Attachment ID.
+	 * @return array{file?: string, filesize: int} Attachment metadata array, with "filesize" value added.
+	 */
+	function set_filesize_in_attachment_meta( array $metadata, int $attachment_id ) : array {
+		$file = get_attached_file( $attachment_id );
+
+		if ( file_exists( $file ) ) {
```
Comment (joehoyle) [inline inc/class-plugin.php:419]: I think this should return early if filesize is already set too, that will mean it's future compatible when images add it.

File: inc/class-plugin.php [hunk 3]
```diff
@@ -396,6 +397,32 @@ public function filter_sideload_move_temp_file_to_s3( array $file ) {
...
+	 * result, operations like `wp_prepare_attachments_for_js' take substantially
+	 * longer to complete against s3 uploads than if they were performed with a
+	 * local filesystem.i
+	 *
+	 * Saving the filesize in the attachment metadata when the image is
+	 * uploaded allows core to skip this stat when retrieving and formatting it.
+	 *
+	 * @param array{file?: string} $metadata      Attachment metadata.
+	 * @param int                  $attachment_id Attachment ID.
+	 * @return array{file?: string, filesize: int} Attachment metadata array, with "filesize" value added.
```
Comment (joehoyle) [inline inc/class-plugin.php:414]: This should also be `filesize?` as it's possible the array won't contain it (if the file doesn't exist). I'd have thought Psalm would have caught that 🤔

**Review by joehoyle — CHANGES_REQUESTED** [review]

**Comment by goldenapples** [comment]
@joehoyle Could you help me understand why the build isn't passing here?

The error is 
```
Running Psalm...

Problem parsing /code/psalm.xml:
  Error on line 2:
    Element '{https://getpsalm.org/schema/config}psalm', attribute 'requireVoidReturnType': The attribute 'requireVoidReturnType' is not allowed.
``` 
but I'm not familiar with what that means, and I think I'm following the style from the documentation and elsewhere in the file.

File: inc/class-plugin.php (hunk 2 above)
Comment (goldenapples) [reply inc/class-plugin.php:419]: Yeah, I was thinking that too. Adding it now.

File: inc/class-plugin.php (hunk 3 above)
Comment (goldenapples) [reply inc/class-plugin.php]: Makes sense, but that didn't seem to fix the build either :thinking:

**Comment by joehoyle** [comment]
Ahh doh it seemed I inadvertently switched to a different version of Psalm ( ` : void` on functions that don't return anything is now required). Will fix up.

File: inc/class-plugin.php (hunk 2 above)
Comment (goldenapples) [reply inc/class-plugin.php]: Fixed in 4c50bf1

File: inc/class-plugin.php [hunk 4]
```diff
@@ -396,6 +397,34 @@ public function filter_sideload_move_temp_file_to_s3( array $file ) {
...
+	 * Saving the filesize in the attachment metadata when the image is
+	 * uploaded allows core to skip this stat when retrieving and formatting it.
+	 *
+	 * @param array{file?: string} $metadata      Attachment metadata.
+	 * @param int                  $attachment_id Attachment ID.
+	 * @return array{file?: string, filesize?: int} Attachment metadata array, with "filesize" value added.
+	 */
+	function set_filesize_in_attachment_meta( array $metadata, int $attachment_id ) : array {
+		$file = get_attached_file( $attachment_id );
+		if ( ! $file ) {
```
Comment (joehoyle) [inline inc/class-plugin.php:418]: @goldenapples `get_attached_file` _can_ return false, so I added the check here -- Psalm was complaining about that. Yey for type safety!

**Review by joehoyle — APPROVED** [review]

**Pull request merged**